### ISSUE TITLE
[C#] Implement new Concept Exercise: pattern-matching-tuples

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/csharp#1443]
This issue describes how to implement the `pattern-matching-tuples` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student some of the advanced features of the Concept of Tuples in C#.

## Learning objectives

- Know how to pattern match on tuples.
- Know how to add tuple pattern matching to your own and built-in types.
- Know that the `ValueTuple` type has obsoleted the `Tuple` type and why.

## Out of scope

Nothing.

## Concepts

This Concepts Exercise's Concepts are:

- `pattern-matching-tuples`: know how to pattern matching on tuples; know how to add tuple pattern matching to other types; know that the `ValueTuple` type has obsoleted the `Tuple` type and why.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `tuples`: know the basics about tuples.
- `pattern-matching-constants`: know how to do pattern matching on constant values.
- `extension-methods`: a built-in library class will have tuple deconstruction added to it.

### Hints

- [Tuple pattern][tuple-pattern]: shows how to pattern match tuples.
- [Deconstructing tuples][deconstructing-tuples]: shows how to do tuple deconstruction for custom and built-in types.

### After

- [Tuple pattern][tuple-pattern]: shows how to pattern match tuples.
- [Deconstructing tuples][deconstructing-tuples]: shows how to do tuple deconstruction for custom and built-in types.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/pattern_matching.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[tuple-pattern]: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#tuple-patterns
[deconstructing-tuples]: https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct
